Source/Core/VideoCommon/Assets/MeshAsset.cpp

```diff
@@ -0,0 +1,597 @@
+            asset_id, material_name);
+        return false;
+      }
+
+      data->m_mesh_material_to_material_asset_id[material_name] = asset_id_json.to_str();
+    }
+  }
+  return true;
+}
+
+void MeshData::ToJson(picojson::object* obj, const MeshData& data)
+{
+  if (!obj) [[unlikely]]
+    return;
+
+  auto& json_obj = *obj;
+
+  picojson::object material_mapping;
+  for (const auto& [material_name, asset_id] : data.m_mesh_material_to_material_asset_id)
+  {
+    material_mapping.emplace(material_name, asset_id);
+  }
+  json_obj.emplace("material_mapping", material_mapping);
+}
+
+bool MeshData::FromDolphinMesh(std::span<const u8> raw_data, MeshData* data)
+{
+  std::size_t offset = 0;
+
+  std::size_t chunk_size = 0;
+  std::memcpy(&chunk_size, raw_data.data(), sizeof(std::size_t));
+  offset += sizeof(std::size_t);
+
+  data->m_mesh_chunks.reserve(chunk_size);
+  for (std::size_t i = 0; i < chunk_size; i++)
+  {
+    MeshDataChunk chunk;
+
+    std::memcpy(&chunk.num_vertices, raw_data.data() + offset, sizeof(u32));
+    offset += sizeof(u32);
+
+    std::memcpy(&chunk.vertex_stride, raw_data.data() + offset, sizeof(u32));
+    offset += sizeof(u32);
+
+    chunk.vertex_data = std::make_unique<u8[]>(chunk.num_vertices * chunk.vertex_stride);
+    std::memcpy(chunk.vertex_data.get(), raw_data.data() + offset,
+                chunk.num_vertices * chunk.vertex_stride);
+    offset += chunk.num_vertices * chunk.vertex_stride;
+
+    std::memcpy(&chunk.num_indices, raw_data.data() + offset, sizeof(u32));
+    offset += sizeof(u32);
+
+    chunk.indices = std::make_unique<u16[]>(chunk.num_indices);
+    std::memcpy(chunk.indices.get(), raw_data.data() + offset, chunk.num_indices * sizeof(u16));
+    offset += chunk.num_indices * sizeof(u16);
+
+    std::memcpy(&chunk.vertex_declaration, raw_data.data() + offset,
+                sizeof(PortableVertexDeclaration));
+    offset += sizeof(PortableVertexDeclaration);
+
+    std::memcpy(&chunk.primitive_type, raw_data.data() + offset, sizeof(PrimitiveType));
+    offset += sizeof(PrimitiveType);
+
+    std::memcpy(&chunk.components_available, raw_data.data() + offset, sizeof(u32));
+    offset += sizeof(u32);
+
+    std::memcpy(&chunk.transform.data[0], raw_data.data() + offset,
+                chunk.transform.data.size() * sizeof(float));
+    offset += chunk.transform.data.size() * sizeof(float);
+
+    std::size_t material_name_size = 0;
+    std::memcpy(&material_name_size, raw_data.data() + offset, sizeof(std::size_t));
+    offset += sizeof(std::size_t);
+
+    chunk.material_name.assign(raw_data.data() + offset,
+                               raw_data.data() + offset + material_name_size);
+    offset += material_name_size * sizeof(char);
+
+    data->m_mesh_chunks.push_back(std::move(chunk));
+  }
+
+  return true;
+}
+
+void MeshData::ToDolphinMesh(File::IOFile* file_data, const MeshData& data)
```

Comment:
Should this not return if the write completely succeeded? What happens if any of the writes fail.